cmake/configs/modes.cmake

```diff
@@ -9,104 +9,173 @@
+    if(DEFINE_BUILD_C_COMPILER_FLAGS)
+        check_c_compiler_flag("${DEFINE_BUILD_C_COMPILER_FLAGS}" HAVE_BUILD_MODE_C_FLAGS)
+        if(NOT HAVE_BUILD_MODE_C_FLAGS)
+            message(VERBOSE "Skipping build mode definition due to unavailable C flags: ${mode}")
+            return()
+        endif()
+    endif()
+    if(DEFINE_BUILD_CXX_COMPILER_FLAGS)
+        check_cxx_compiler_flag("${DEFINE_BUILD_CXX_COMPILER_FLAGS}" HAVE_BUILD_MODE_CXX_FLAGS)
+        if(NOT HAVE_BUILD_MODE_CXX_FLAGS)
+            message(VERBOSE "Skipping build mode definition due to unavailable CXX flags: ${mode}")
+            return()
+        endif()
+    endif()
+    unset(CMAKE_REQUIRED_FLAGS)
+    unset(HAVE_BUILD_MODE_C_FLAGS)
+    unset(HAVE_BUILD_MODE_CXX_FLAGS)
+
+    string(REPLACE ";" " " DEFINE_BUILD_C_COMPILER_FLAGS "${DEFINE_BUILD_C_COMPILER_FLAGS}")
+    string(REPLACE ";" " " DEFINE_BUILD_CXX_COMPILER_FLAGS "${DEFINE_BUILD_CXX_COMPILER_FLAGS}")
+    string(REPLACE ";" " " DEFINE_BUILD_LINK_FLAGS "${DEFINE_BUILD_LINK_FLAGS}")
+    string(TOUPPER ${mode} build_mode)
+    set(CMAKE_C_FLAGS_${build_mode}
+        "${DEFINE_BUILD_C_COMPILER_FLAGS}" CACHE STRING
+        "Flags used by the C compiler for ${mode} build type or configuration." FORCE)
+
+    set(CMAKE_CXX_FLAGS_${build_mode}
+        "${DEFINE_BUILD_CXX_COMPILER_FLAGS}" CACHE STRING
+        "Flags used by the C++ compiler for ${mode} build type or configuration." FORCE)
+
+    set(CMAKE_EXE_LINKER_FLAGS_${build_mode}
+        "${DEFINE_BUILD_LINK_FLAGS}" CACHE STRING
+        "Linker flags to be used to create executables for ${mode} build type." FORCE)
+
+    set(CMAKE_SHARED_LINKER_FLAGS_${build_mode}
+        "${DEFINE_BUILD_LINK_FLAGS}" CACHE STRING
+        "Linker lags to be used to create shared libraries for ${mode} build type." FORCE)
 
     mark_as_advanced(
-        CMAKE_CXX_FLAGS_ASAN
-        CMAKE_C_FLAGS_ASAN
-        CMAKE_EXE_LINKER_FLAGS_ASAN
-        CMAKE_SHARED_LINKER_FLAGS_ASAN
+        CMAKE_CXX_FLAGS_${build_mode}
+        CMAKE_C_FLAGS_${build_mode}
+        CMAKE_EXE_LINKER_FLAGS_${build_mode}
+        CMAKE_SHARED_LINKER_FLAGS_${build_mode}
     )
-    list(APPEND build_modes "ASan")
-endif()
+    set(BUILD_MODES "${BUILD_MODES};${mode}" CACHE INTERNAL "")
+endfunction()
 
-# Create an UBSAN build variant
-if("${CMAKE_C_COMPILER_ID}" STREQUAL "MSVC")
-    set(ubsan_link_flags "/fsanitize=undefined")
-    set(ubsan_compiler_flag "/fsanitize=undefined")
-elseif("${CMAKE_C_COMPILER_ID}" STREQUAL "Clang" OR "${CMAKE_C_COMPILER_ID}" STREQUAL "AppleClang")
-    set(ubsan_link_flags "-fsanitize=undefined")
-    set(ubsan_compiler_flag "-fsanitize=undefined")
+if(MSVC)
+    set(no_omit_frame_flag "/Oy-")
 else()
-    set(ubsan_link_flags "-fsanitize=undefined -lubsan")
-    set(ubsan_compiler_flag "-fsanitize=undefined")
+    set(no_omit_frame_flag "-fno-omit-frame-pointer")
+    set(no_omit_frame_flag "-fno-omit-frame-pointer")
```

Comment:
I think this line is a duplicate of the previous one.

Reply:
Good find! Fixed 05ba22b5f134c16ed84cc98924cbee02a119bf9c